design/specs/visual-query-builder.md

```diff
@@ -0,0 +1,140 @@
+---
+title: Visual Query Builder
+description: 
+published: true
+date: 2022-03-08 18:02:53
+tags: 
+editor: markdown
+dateCreated: 2022-03-08 18:02:58
+---
+
+## Context
+
+This document specifies the design for a visual query builder named `Data Explorer` within the Mathesar application. Through `Data Explorer`, users will be able to analyze data across one or multiple tables of a schema and save the resulting tables as views.
+
+Additionally, users will be able to open existing views in `Data Explorer`, as long as its query commands and functions are supported.
+
+## Scenarios
+
+## 1. Selecting the Base Table
+
+The base table selection is the first step when starting a new query, and it will determine which columns and links are available for selection.
+
+The base table also determines the automatic joins that are performed when the user adds columns from linked tables.
+
+### Changing the base table
+
+Once the input table has been created, changing the base table will result in the loss of all progress.
+
+### 1.1 Base Table Options
+
+The base table selector will display a list of all available tables in the current schema. Initially, this list will not include views which will be added in later iterations.
+
+To select a table, the user will find it on the list and click on the desired table option.
+
+Selecting the base table will enable the interface controls for adding additional steps, which were initially disabled. No action is available before a base table is selected.
+
+Once a selection has been made, the user will be prompted to proceed with the input table configuration, starting with column selection.
+
+[BASE TABLE SELECT WIREFRAME]
+
+## 2. Creating the Input Table
+
+The input table is created by selecting columns from the base table or creating new ones via formulas.
+
+Various options are available for input columns, including filters or aggregations that can help users control which data they want to show and how.
+
+These options can be set for all input columns, including formulas. Some options like aggregations might only be available for columns that link to multiple values. The system will automatically detect this and present the available options to users.
```

Comment:
I think you should specify that filters and aggregations specifically will only be available for columns that involve pulling data from multiple records. It's clearer when implementing.

It might be good to document all of the options available for a column and rules for when they are available.